src/parser.cc

```diff
@@ -8,35 +8,78 @@
 #include "xcshell/constants.h"
 #include "xcshell/utils.h"
 
-CommandParseResult buildParseResultWithRedirect(
+bool Parser::IsRedirect(const std::string &command_with_arg) {
+  if (command_with_arg == REDIRECT_OUTPUT_OVERWRITE ||
+      command_with_arg == REDIRECT_INPUT ||
+      command_with_arg == REDIRECT_OUTPUT_APPEND ||
+      command_with_arg == REDIRECT_ERROR_OUTPUT_OVERWRITE ||
+      command_with_arg == REDIRECT_ERROR_OUTPUT_APPEND) {
+    return true;
+  }
+  return false;
+}
+
+bool Parser::IsErrorRedirect(const std::string &command_with_arg) {
+  if (command_with_arg == REDIRECT_ERROR_OUTPUT_APPEND ||
+      command_with_arg == REDIRECT_ERROR_OUTPUT_OVERWRITE ||
+      command_with_arg == REDIRECT_ERROR_AND_OUTPUT) {
+    return true;
+  }
+  return false;
+}
+
+bool Parser::IsRedirectOverwrite(const std::string &command_with_arg) {
+  if (command_with_arg == REDIRECT_ERROR_OUTPUT_APPEND ||
+      command_with_arg == REDIRECT_OUTPUT_APPEND) {
+    return true;
+  }
+  return false;
+}
+
+bool Parser::IsOutputRedirectSymbol(const std::string &command_with_arg) {
+  if (command_with_arg == REDIRECT_OUTPUT_OVERWRITE ||
+      command_with_arg == REDIRECT_OUTPUT_APPEND ||
+      command_with_arg == REDIRECT_ERROR_OUTPUT_OVERWRITE ||
+      command_with_arg == REDIRECT_ERROR_OUTPUT_APPEND) {
+    return true;
+  }
+  return false;
+}
+
+bool Parser::IsInputRedirectSymbol(const std::string &command_with_arg) {
+  if (command_with_arg == REDIRECT_INPUT) {
+    return true;
+  }
+  return false;
+}
+
+CommandParseResult Parser::BuildParseResultWithRedirect(
     const std::vector<std::string> &command_with_args,
     const std::string &command) {
   std::vector<std::string> args;
   bool args_end = false;
   std::string output_file;
   std::string input_file;
-  bool is_overwrite;
+  bool is_overwrite = false;
+  bool is_error_redirect = false;
   for (int i = 0; i < command_with_args.size(); i++) {
     auto command_with_arg = command_with_args[i];
-    if (command_with_arg == REDIRECT_OUTPUT_OVERWRITE ||
-        command_with_arg == REDIRECT_INPUT ||
-        command_with_arg == REDIRECT_OUTPUT_APPEND) {
-      args_end = true;
-    }
+    if (IsRedirect(command_with_arg)) args_end = true;
     if (!args_end) {
       args.push_back(command_with_arg);
     } else if ((i + 1) < command_with_args.size()) {
-      if (command_with_arg == REDIRECT_OUTPUT_OVERWRITE ||
-          command_with_arg == REDIRECT_OUTPUT_APPEND) {
+      if (IsOutputRedirectSymbol(command_with_arg)) {
         output_file = command_with_args[i + 1];
```

Comment:
Stdout and stderr should able to have different file, don't mix them together. See the example below.
![WX20211009-110859](https://user-images.githubusercontent.com/14891023/136642212-21c40d7e-03f4-450b-b47e-8b82262fe96a.png)

![WX20211009-111205](https://user-images.githubusercontent.com/14891023/136642214-8557898c-fb90-4b61-9922-b951a8803d48.png)
.